### DISCUSSION TITLE
virtual branches feature, like GitButler

### DISCUSSION THREAD

**Discussion by aspiers** [magit/magit#5073] [Ideas]
I just discovered GitButler, and it looks like a very promising solution to the problem of dealing with multiple concurrent threads of work:

https://docs.gitbutler.com/features/virtual-branches

I recommend watching the short YouTube demo on the above page.

I wrote some [ugly shell hacks](https://github.com/aspiers/git-config/tree/master?tab=readme-ov-file#management-of-multiple-concurrent-branches) for this years ago which work OK for me, but I've longed for a better solution to these higher-order workflows.  It would be amazing if at some magical point in the future, magit could achieve the same kind of thing.

**Comment by xxyyz** [comment]
I wanted something similar as well. I think a (much) simpler variant could still be very useful.

What about having the possibility to categorize (mark / tag) hunks (in the work tree and/or in the index). So then the usual commands (stage/commit/...) could target the given categories.

The work tree and the index can be seen as two categories of "coming changes". And the applicable commands (stage/commit...) are restricted in each category. It would be great to extend this idea and allow the creation of any number of categories, ideally with user defined names (but this is not essential). And allow all(?) commands to apply to any category.

E.g. I have 10 hunks in the work tree. When I check them, currently I can stage any of them and then e.g. commit or stash the selected ones, i.e. the ones, that are "moved" into the index (or kept in the work tree). This is a possibility to split bigger changes into smaller commits, but it is a tedious task when there are many hunks and I want to organize them into several commits. The idea is to mark the hunks; in other words: to assign them to categories X or Y or Z or ... Then, having commands like "stash the hunks in category X" or "commit the hunks in category Y" or "create a new branch named Z and commit the changes in category Z in that new branch"...

As I understand, "virtual branches" may provide one more feature on top of these: they can be used to separate the changes in time, so if I move a hunk into a virtual branch (into a category), and then change the corresponding code again (but still before committing the original changes), these subsequent changes can easily be separated (into different "virtual branches"). With "normal git" (and thus: with the simple hunk-categorization feature sketched above), this separation is supported only in the two predefined categories, index and work tree, but e.g. separating three subsequent changes of the same code is not possible (without stashing or committing... meanwhile). In my view this extra functionality of "virtual branches" is a nice add on, but its implementation may be complicated.

I hope that the implementation of the hunk categorization feature sketched above, would not be difficult in magit.

**Comment by tarsius** [comment]
This is a nice feature, but I expect it would be very hard to implement properly. An initial poc *may* turn out to be not *that* hard, but if I remember correctly, the video only shows the happy path, leaving out what happens if there are conflicts between the things being committed to different branches. Getting it to work properly in all cases and to provide adequate support when things don't just work and the user has to get involved, would definitely be hard, and would require making improvements over the next few years.

I simply do not have the time for that. Sorry. I don't want to implement a 80% version, because I know that I would then be on the hook to do the other 80%. I do not need another time sink. This is their killer feature, and they have a team working on it. I cannot do that also, on top over everything else.

What I have been using instead is separate *commits*, which may, in a second, separate step be split across different branches. Changes can be assigned to different buckets using "instant fixup". Magit also provides useful tools for splitting, such as `magit-cherry-harvest`. Using that command can lead to the user having to resolve conflicts, but they happen during rebase (or even the initial apply), so they are well understood.

Maybe that's just because that's what I am used to, but I think this might actually be superior. I think there is value in separating the steps of "assigning" changes to "buckets" (either commits or branches), and performing the final separation in a separate step. (Or to even do several rounds of it.)

I don't know what you would want to use this feature for. I could see myself using it when I work on some feature and spot an issue nearby. A trivial example would be a typo in nearby function that won't also be touched by the changes I am actually working on. In that case one can just create a new commit, (optionally) rebase it toward the beginning of the feature branch (to move it "out of sight"), and later extract it onto another branch, using the harvest command.

A much less trivial case is, when you realize that, what you have in the working tree, is not just a new feature, but also a refactoring of the existing code on top of which the new feature is being implemented. This happens to me a lot and in many cases it is simply impossible to separate the refactored part from the new parts, using just `git add`, `git apply` and friends, or even tooling on top of those.

I do it anyway, by committing the combined result, doing a rebase that stops at the commit before that "combined" commit, and then recreating the refactoring parts by themselves (potentially spit across multiple commits). (Some parts I might be able to apply from the "next"/"combined" commit, but some manual editing is involved, in the hard cases at least). Then I continue the rebase, which results in a conflict, at which point I can just pick "their" side. The diff after that is done has to be inspected carefully. It is possible that one included too little or too much in the (now) "earlier" commit.

This isn't trivial, but I think there is value in doing it even in the cases that require "re-editing". The multiple-branches approach wouldn't magically make the hard cases trivial. I have a feeling that it would make those harder or even impossible. The value of the rebase approach is in having to rethink the steps separately, again, or if the work-process was a bit muddy, actually for the first time. It does definitely happen that in this process, I don't just come up with cleaner steps toward the end result, but actually a cleaner result, including a cleaner new feature *and*/or a cleaner base abstraction.

I guess one could say I think about this "in time" (series of commits, some of which involves time travel), while you think about it "in space" (branches that co-exist). The solution in space is probably more intuitive, but it only works in the trivial cases. In my experience, the non-trivial cases are actually quite common. Making the trivial cases even easier to handle, doesn't seem like a good investment for me. Furthermore, using the time-traveling approach works here too, and using that for the trivial cases, has the advantage that one gets the practice, to later deal with the non-trivial cases with confidence. I don't think the non-trivial cases can be done in space alone.

**Reply by xxyyz** [reply]
Thank you very much for your quick answer and detailed explanation. Using instant fixup for sorting the hunks "instantly" into different commits, is a good idea, I'll try it out next time. As you can see below, I target the same issue, in another way. My proposal could make it even more easier.

Since I am not sure if my proposal went through, so let me rephrase. I think my proposal covers 80% of "virtual branches" of GitButler, without "virtual branches" and without all the complications it involves. But my proposal can be implemented in 10 (1?) % of the time needed for the implementation of  virtual branches, I guess. (Maybe I should have written it in a separate "ideas" thread...)

Let's assume I have 10 hunks, all are unstaged, nothing is staged yet. When I think it is time to commit them, at least partially, I look through my changes, i.e. I check each hunk one by one. I know that they fall in three categories, e.g. "1) improving performance of function x; 2) add new functionality y and 3) fix "nearby" typos/docs/formatting...". When I started the work, I wanted 2, but meanwhile I found 1 and 3 accidentally and couldn't resist to implement them.

Nowadays, in this case, I do the following: stage the changes for 1, then commit them. Then stage the changes for 2, then commit them. Then stage the changes for 3, then commit them. For this, I have to check each hunk at the beginning (to find the ones belonging to 1). After committing 1, I have to _re_check all the remaining hunks to find the ones belonging to 2. So I have to check some hunks twice. This is not a problem if I have not too many hunks and not too many categories (commits, at the end). But if I would have 50 or 100 or 300 hunks (think e.g. a rename of a function or variable used in many places) and 5 commits, then I would have to recheck many hunks 4 times, just to leave them in the work tree, before finally I stage them for the last commit. My proposal would ease this sorting of the hunks. (Using instant fixup, as you suggest, would eliminate the need to check the same hunk several times. But then I have to commit (fixup) many times.)

I think GitButler's virtual branches does "almost" the same, with 1 "little" extra twist, which makes it slightly more useful for the users, but much harder to implement. My proposal, on the other hand, can be implemented much easier, if I am not mistaken.

My proposal is to add (besides the s(tage) command) a "mark" command to the hunks. Mark could take an argument (category name: '1', '2' and '3' in the example above). This "marking", could be done "above" git level: git do not have to know anything about it, everything could be done in magit: magit should know which hunk is assigned to which category. When I marked all hunks I wanted (in the example above: I assigned all the hunks to category 1, 2 or 3), I could say "stage (or: commit) all hunks marked '1'". Then do the same for the hunks in category 2 and finally for 3. This way I have to check all hunks only once.

If a hunk is changed by a subsequent code change, its mark could simply disappear automatically (when I r(efresh) the magit buffer). (It is another change, after all, than the one marked originally.) GitButler's "virtual branches" may be prepared to handle such conflicts, which makes its implementation complicated, I guess.

Maybe this (marking) functionality could be limited to the work tree initially?

As you see, no branching is involved here (unless we would have a command to "commit the hunks marked x in a new branch named x", but that just seems to be a simple automation of the sequence of a few already existing steps: commit the hunks, and spin off the last commit to that new branch...)

Would this still be problematic in magit? Or you are not convinced that it would be useful? Honestly I "need" it rarely, but sometimes it would be very handy. Or is there a simple way to do this already? (I'll try the suggested instant fixup method!)

**Reply by tarsius** [reply]
I don't really have the time to implement, design or even just review such a feature. A minimal viable version would definitely lead to follow up requests, *especially* if it works well (except when it doesn't).

It's also not like I haven't though about this before. I haven't so far in me deciding it's a bad idea either. It's just that there are a lot of other things I have or want to do, including other exciting new things, which I am more confident are feasible and/or have greater potential. So this is something that's on the back of my mind, and it has been for a while, but it is far from a priority. Of course I could just do a "quick experiment". But that's also true for some of those other things. I am currently trying to avoid, getting sucked into another stream of "just one more thing".

> This "marking", could be done "above" git level:

Magit has no support for such marking. Implementing it for hunks is more difficult than for any other section type (at least I cannot think of a more difficult type on the spot). Hunks change their "identity" when other hunks appear or disappear. We could hash the contents of hunks to arrive at a more permanent identity.

So an initial poc would be easier to do at the "git level".

I think these two commands would do:

1. `create-buckets` Read a series of short descriptions from the user. When done, use each to create an empty commit.
2. `assign-to-buckets` (Maybe bail if there are staged changes.) Call `magit-stage`. Make the user select one of the buckets (using e.g., `completing-read` or a dynamically populate transient menu). Create a fixup commit for the selected bucket commit, with the staged changes. (This could alternatively just let the user select any non-fixup commit that isn't in the upstream branch, removing the need for `create-bucket`.)
